instrumentation/jms/src/main/java/brave/jms/JmsAdapter.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2013-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.jms;
+
+import brave.messaging.MessagingAdapter;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.Message;
+import javax.jms.Queue;
+import javax.jms.Topic;
+
+abstract class JmsAdapter<T> extends MessagingAdapter<Destination, T, T> {
+  final String remoteServiceName;
+
+  JmsAdapter(JmsTracing jmsTracing) {
+    remoteServiceName = jmsTracing.remoteServiceName;
+  }
+
+  static final class MessageAdapter extends JmsAdapter<Message> {
+    MessageAdapter(JmsTracing jmsTracing) {
+      super(jmsTracing);
+    }
+
+    @Override public String correlationId(Message message) {
+      try {
+        return message.getJMSMessageID();
+      } catch (JMSException ignored) {
+        // don't crash on wonky exceptions!
+      }
+      return null;
+    }
+  }
+
+  @Override public T carrier(T message) {
+    return message;
+  }
+
+  @Override public String channel(Destination channel) {
+    try {
+      if (channel instanceof Queue) {
+        return ((Queue) channel).getQueueName();
+      } else if (channel instanceof Topic) {
+        return ((Topic) channel).getTopicName();
+      }
+      // TODO: we could use toString here..
```

Comment:
If we follow `destination.toString()` instead of casting, we might get something like `ActiveMQQueue[queueName]`. I'm not against but it could make difficult to process/aggregate later. I'm thinking about #895  

OTOH, we could create a `Channel` data type, that could allow use to do the casting only once. WDYT?

Reply:
oh it wasn't about the above casting, it was about when we don't know the type.. for example what to do then.

I'm worried about another type as our generic complexity is already a lot.